addons/mail/models/mail_scheduled_message.py

```diff
@@ -0,0 +1,262 @@
+
+from collections import defaultdict
+import json
+
+from odoo import _, api, fields, models
+from odoo.addons.mail.tools.discuss import Store
+from odoo.exceptions import AccessError, UserError, ValidationError
+
+import logging
+
+_logger = logging.getLogger(__name__)
+
+
+class ScheduledMessage(models.Model):
+    """ Scheduled message model (holds post values generated by the composer to delay the
+    posting of the message). Different from mail.message.schedule that posts the message but
+    delays the notification process.
+
+    Todo: when adding support for scheduling messages in mass_mail mode, could add a reference to
+    the "parent" composer (by making 'mail.compose.message' not transient anymore). This reference
+    could then be used to cancel every message scheduled "at the same time" (from one composer),
+    and to get the static 'notification parameters' (mail_server_id, auto_delete,...) instead of
+    duplicating them for each scheduled message.
+    Currently as scheduling is allowed in monocomment only, we don't have duplicates and we only
+    have static notification parameters, but some will become dynamic when adding mass_mail support
+    such as 'email_from' and 'force_email_lang'.
+    """
+    _name = 'mail.scheduled.message'
+    _description = 'Scheduled Message'
+
+    # content
+    subject = fields.Char('Subject')
+    body = fields.Html('Contents', sanitize_style=True)
+    scheduled_date = fields.Datetime('Scheduled Date', required=True)
+    attachment_ids = fields.Many2many(
+        'ir.attachment', 'scheduled_message_attachment_rel',
+        'scheduled_message_id', 'attachment_id',
+        string='Attachments')
+
+    # related document
+    model = fields.Char('Related Document Model', required=True)
+    res_id = fields.Many2oneReference('Related Document Id', model_field='model', required=True)
+
+    # origin
+    author_id = fields.Many2one('res.partner', 'Author', required=True)
+
+    # recipients
+    partner_ids = fields.Many2many('res.partner', string='Recipients')
+
+    # characteristics
+    is_note = fields.Boolean('Is a note', default=False, help="If the message will be posted as a Note.")
+
+    # notify parameters (email_from, mail_server_id, force_email_lang,...)
+    notification_parameters = fields.Text('Notification parameters')
+
+    @api.constrains('scheduled_date')
+    def _check_scheduled_date(self):
+        if any(scheduled_message.scheduled_date < fields.Datetime().now() for scheduled_message in self):
+            raise ValidationError(_("A Scheduled Message cannot be scheduled in the past"))
+
+    # ------------------------------------------------------
+    # CRUD / ORM
+    # ------------------------------------------------------
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        # check that messages can be posted on the models
+        for model in {vals['model'] for vals in vals_list}:
```

Comment:
Would explicitly remove default_model and default_res_id (we know what sneaky monkeys try to do), so that we are sure model and res_id are given in create (or it is going to raise for missing values).